docs/layout-of-source-files.rst

```diff
@@ -216,101 +216,562 @@ the code below creates new global symbols ``alias`` and ``symbol2`` which refere
-If multiple remappings apply, the one with the longest key is tried first.
-An empty prefix is not allowed. The remappings can depend on a context,
-which allows you to configure packages to import e.g., different versions of a
-library of the same name.
+    import "../util.sol";          // Import key: /a/b/util.sol
+    import "../../util.sol";       // Import key: /a/util.sol
+    import "../../../util.sol";    // Import key: /util.sol
 
-**solc**:
+    import "../././.././util.sol"; // Import key: /a/util.sol
 
-For solc (the commandline compiler), you provide these path remappings as
-``context:prefix=target`` arguments, where both the ``context:`` and the
-``=target`` parts are optional (``target`` defaults to ``prefix`` in this
-case). All remapping values that are regular files are compiled (including
-their dependencies).
+If there are more ``../`` segments than directory segments in the parent import key, the evaluation
+stops at the root:
 
-This mechanism is backwards-compatible (as long
-as no filename contains ``=`` or ``:``) and thus not a breaking change. All
-files in or below the ``context`` directory that import a file that starts with
-``prefix`` are redirected by replacing ``prefix`` by ``target``.
+.. code-block:: solidity
+    :caption: /a/b/c/contract.sol
 
-For example, if you clone ``github.com/ethereum/dapp-bin/`` locally to
-``/usr/local/dapp-bin``, you can use the following in your source file:
+    import "../../../../util.sol";       // Import key: /util.sol
+    import "../../../../../util.sol";    // Import key: /util.sol
+    import "../../../../../../util.sol"; // Import key: /util.sol
+
+.. code-block:: solidity
+    :caption: a/b/c/contract.sol
+
+    import "../../../../util.sol";       // Import key: util.sol
+    import "../../../../../util.sol";    // Import key: util.sol
+    import "../../../../../../util.sol"; // Import key: util.sol
+
+After stripping the leading relative segments, the import path is normalized so that the
+resulting import key does not contain any ``./`` or ``../``:
+
+.. code-block:: solidity
+    :caption: /a/b/c/contract.sol
+
+    import "../../d/e///.././util.sol"; // Import key: /a/e/util.sol
+
+This is quite different from imports relative to base where the ``///.././`` part would remain
+in the import key.
+
+Note that the parent import key is **not** normalized, and the ``./`` and ``../`` segments in it
+have no special meaning:
+
+.. code-block:: solidity
+    :caption: ../lib/math.sol
+
+    import "./util.sol" as util;    // Import key: ../lib/util.sol
+    import "../token.sol" as token; // Import key: ../../token.sol
+
+This may lead to surprising results in corner cases:
+
+.. code-block:: solidity
+    :caption: /a/./b/contract.sol
+
+    import "../c/util.sol";       // Import key: /a/./c/util.sol
+    import "../../c/util.sol";    // Import key: /a/c/util.sol
+    import "../../../c/util.sol"; // Import key: /c/util.sol
```

Comment:
This is how it works currently but I think it would be more intuitive if we changed it so that the whole resulting path is normalized.

The only reason I can think of for this behavior might be not to break something like this:
```json
{
   "language": "Solidity",
   "sources": {
       "https://github.com/user/repo/contract.sol": {
           "content": "import \"./token.sol\";\ncontract C {}"
       },
       "https://github.com/user/repo/token.sol": {
           "content": "library Util {}"
       }
   },
   "settings": {
       "outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}
   }
}
```

With parent key normalization `./token.sol` would resolve to `https:/github.com/user/repo/token.sol` (note the missing slash).

But I think that we might be better off just adding a special case for URLs and treating everything else as paths.